File: src/main/java/com/breakreasi/voip_android_2/rests/Rests.java
package com.breakreasi.voip_android_2.rests;

import android.content.Context;

import androidx.annotation.NonNull;

import com.breakreasi.voip_android_2.rests.ResponseInitConfigModel;
import com.breakreasi.voip_android_2.rests.ResponseModel;
import com.breakreasi.voip_android_2.rests.RestsRepo;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.net.MalformedURLException;
import java.net.URL;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class Rests {
    private Retrofit retrofit;
    private final RestsRepo repose;
    private final String API_KEY = "1234";

    public Rests() {
        Retrofit client = getClient("https://api-receiver.jasvicall.my.id/");
        repose = client.create(RestsRepo.class);
    }

    public Retrofit getClient(String baseUrl) {
        Gson gson = new GsonBuilder()
                .setLenient()
                .create();

        URL myURL = null;
        try {
            myURL = new URL(baseUrl);
        } catch (MalformedURLException ignored) {
        }
        if (retrofit == null) {
            if (myURL != null) {
                retrofit = new Retrofit.Builder()
                        .baseUrl(myURL)
                        .addConverterFactory(GsonConverterFactory.create(gson))
                        .build();
            }
        } else {
            if (myURL != null) {
                if (!retrofit.baseUrl().equals(myURL)) {
                    retrofit = new Retrofit.Builder()
                            .baseUrl(myURL)
                            .addConverterFactory(GsonConverterFactory.create(gson))
                            .build();
                }
            }
        }
        return retrofit;
    }

    public void initConfig(ResponseCallback<ResponseInitConfigModel> callback) {
        repose.initConfig(API_KEY).enqueue(new Callback<>() {
            @Override
            public void onResponse(@NonNull Call<ResponseInitConfigModel> call, @NonNull Response<ResponseInitConfigModel> response) {
                if (response.body() != null && (response.code() == 201 || response.code() == 200)) {
                    callback.onResponse(response.body());
                } else {
                    callback.onResponse(null);
                }
            }
            @Override
            public void onFailure(@NonNull Call<ResponseInitConfigModel> call, @NonNull Throwable t) {
                callback.onResponse(null);
            }
        });
    }

    public void acceptCall(String token, String channel, ResponseCallback<ResponseModel> callback) {
        repose.acceptCall(API_KEY, token, channel).enqueue(new Callback<>() {
            @Override
            public void onResponse(@NonNull Call<ResponseModel> call, @NonNull Response<ResponseModel> response) {
                if (response.body() != null && (response.code() == 201 || response.code() == 200)) {
                    callback.onResponse(response.body());
                } else {
                    callback.onResponse(null);
                }
            }
            @Override
            public void onFailure(@NonNull Call<ResponseModel> call, @NonNull Throwable t) {
                callback.onResponse(null);
            }
        });
    }

    public void rejectCall(String token, ResponseCallback<ResponseModel> callback) {
        repose.rejectCall(API_KEY, token).enqueue(new Callback<>() {
            @Override
            public void onResponse(@NonNull Call<ResponseModel> call, @NonNull Response<ResponseModel> response) {
                if (response.body() != null && (response.code() == 201 || response.code() == 200)) {
                    callback.onResponse(response.body());
                } else {
                    callback.onResponse(null);
                }
            }
            @Override
            public void onFailure(@NonNull Call<ResponseModel> call, @NonNull Throwable t) {
                callback.onResponse(null);
            }
        });
    }

    public void makeCall(String whatsapp, String lokasi, String token, boolean withVideo, String type, ResponseCallback<ReceiverResponse> callback) {
        repose.makeCall(API_KEY, whatsapp, lokasi, token, withVideo, type).enqueue(new Callback<>() {
            @Override
            public void onResponse(@NonNull Call<ReceiverResponse> call, @NonNull Response<ReceiverResponse> response) {
                if (response.body() != null && (response.code() == 201 || response.code() == 200)) {
                    callback.onResponse(response.body());
                } else {
                    callback.onResponse(null);
                }
            }
            @Override
            public void onFailure(@NonNull Call<ReceiverResponse> call, @NonNull Throwable t) {
                callback.onResponse(null);
            }
        });
    }
}

File: src/main/java/com/breakreasi/voip_android_2/rests/Rests.kt
package com.breakreasi.voip_android_2.rests

import com.google.gson.GsonBuilder
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.net.MalformedURLException
import java.net.URL

class Rests {
    private var retrofit: Retrofit? = null
    private val repose: RestsRepo
    private val API_KEY = "1234"

    init {
        val client = getClient("https://api-receiver.jasvicall.my.id/")
        repose = client.create<RestsRepo>(RestsRepo::class.java)
    }

    fun getClient(baseUrl: String?): Retrofit {
        val gson = GsonBuilder()
            .setLenient()
            .create()

        var myURL: URL? = null
        try {
            myURL = URL(baseUrl)
        } catch (ignored: MalformedURLException) {
        }
        if (retrofit == null) {
            if (myURL != null) {
                retrofit = Retrofit.Builder()
                    .baseUrl(myURL)
                    .addConverterFactory(GsonConverterFactory.create(gson))
                    .build()
            }
        } else {
            if (myURL != null) {
                if (retrofit!!.baseUrl() != myURL) {
                    retrofit = Retrofit.Builder()
                        .baseUrl(myURL)
                        .addConverterFactory(GsonConverterFactory.create(gson))
                        .build()
                }
            }
        }
        return retrofit!!
    }

    fun initConfig(callback: ResponseCallback<ResponseInitConfigModel?>) {
        repose.initConfig(API_KEY)!!.enqueue(object : Callback<ResponseInitConfigModel?> {
            override fun onResponse(
                call: Call<ResponseInitConfigModel?>,
                response: Response<ResponseInitConfigModel?>
            ) {
                if (response.body() != null && (response.code() == 201 || response.code() == 200)) {
                    callback.onResponse(response.body())
                } else {
                    callback.onResponse(null)
                }
            }

            override fun onFailure(call: Call<ResponseInitConfigModel?>, t: Throwable) {
                callback.onResponse(null)
            }
        })
    }

    fun acceptCall(token: String?, channel: String?, callback: ResponseCallback<ResponseModel?>) {
        repose.acceptCall(API_KEY, token, channel)!!.enqueue(object : Callback<ResponseModel?> {
            override fun onResponse(
                call: Call<ResponseModel?>,
                response: Response<ResponseModel?>
            ) {
                if (response.body() != null && (response.code() == 201 || response.code() == 200)) {
                    callback.onResponse(response.body())
                } else {
                    callback.onResponse(null)
                }
            }

            override fun onFailure(call: Call<ResponseModel?>, t: Throwable) {
                callback.onResponse(null)
            }
        })
    }

    fun rejectCall(token: String?, callback: ResponseCallback<ResponseModel?>) {
        repose.rejectCall(API_KEY, token)!!.enqueue(object : Callback<ResponseModel?> {
            override fun onResponse(
                call: Call<ResponseModel?>,
                response: Response<ResponseModel?>
            ) {
                if (response.body() != null && (response.code() == 201 || response.code() == 200)) {
                    callback.onResponse(response.body())
                } else {
                    callback.onResponse(null)
                }
            }

            override fun onFailure(call: Call<ResponseModel?>, t: Throwable) {
                callback.onResponse(null)
            }
        })
    }

    fun makeCall(
        whatsapp: String?,
        lokasi: String?,
        token: String?,
        withVideo: Boolean,
        type: String?,
        callback: ResponseCallback<ReceiverResponse?>
    ) {
        repose.makeCall(API_KEY, whatsapp, lokasi, token, withVideo, type)!!
            .enqueue(object : Callback<ReceiverResponse?> {
                override fun onResponse(
                    call: Call<ReceiverResponse?>,
                    response: Response<ReceiverResponse?>
                ) {
                    if (response.body() != null && (response.code() == 201 || response.code() == 200)) {
                        callback.onResponse(response.body())
                    } else {
                        callback.onResponse(null)
                    }
                }

                override fun onFailure(call: Call<ReceiverResponse?>, t: Throwable) {
                    callback.onResponse(null)
                }
            })
    }
}
